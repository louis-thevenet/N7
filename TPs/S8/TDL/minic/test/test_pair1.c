/*
Ce programme vérifie la bonne gestion des Paires (Couples) en MiniC avec des exemples valides
*/

MonProgramme{
    // Déclaration de variables
    int a = 5;
    int b = 10;
    
    // Déclaration de couple
    <int, int> couple1 = <a, b>; // Initialisation avec des variables
    <int, int> couple2 = <20, 42>; // Initialisation avec des valeurs constantes
    <int, int> couple3 = <a+b, b-a>; // Initialisation avec des expressions
    <boolean , int> couple4 = <true, 42>; // Couple avec un booléen et un entier

    print fst(couple4);
    print snd(couple4);

    // Accès aux éléments du couple
    int premier = fst(couple1); // Accès au premier élément
    int second = snd(couple1); // Accès au deuxième élément

    print premier;
    print second;

    // Modification des éléments du couple
    couple1 = <fst(couple1) + 1, snd(couple1) + 2>; // Incrémentation des éléments

    // Utilisation du couple dans une condition
    if (fst(couple1) > 5) {
        couple1 = <fst(couple1) - 1, snd(couple1)>; 
    } else {
        couple1 = <fst(couple1) + 1, snd(couple1)>; 
    }

    
    // Utilisation du couple dans une fonction
    
    <int,int> JamaisPair(<int,int> c) {
        if (fst(c) % 2 == 0 && snd(c) % 2 == 0) {
            return JamaisPair(<fst(c)/2, snd(c)/2>);
        } else {
            return c;
        }

    print JamaisPair(couple2);
    }


}