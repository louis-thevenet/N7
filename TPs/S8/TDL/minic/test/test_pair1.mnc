/*
Ce programme v√©rifie la bonne gestion des Paires (Couples) en MiniC avec des exemples valides
*/