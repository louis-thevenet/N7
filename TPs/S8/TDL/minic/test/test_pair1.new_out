true42510612224476
