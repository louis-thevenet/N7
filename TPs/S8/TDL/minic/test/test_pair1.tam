PUSH 1
LOADL 5
STORE (1) 0[SB]
PUSH 1
LOADL 10
STORE (1) 1[SB]
PUSH 2
LOAD (1) 0[SB]
LOAD (1) 1[SB]
STORE (2) 2[SB]
PUSH 2
LOADL 20
LOADL 42
STORE (2) 4[SB]
PUSH 2
LOAD (1) 0[SB]
LOAD (1) 1[SB]
SUBR IAdd
LOAD (1) 1[SB]
LOAD (1) 0[SB]
SUBR ISub
STORE (2) 6[SB]
PUSH 2
LOADL 1
LOADL 42
STORE (2) 8[SB]
LOAD (2) 8[SB]
POP (0) 1
SUBR BOut
LOAD (2) 8[SB]
POP (1) 1
SUBR IOut
PUSH 1
LOAD (2) 2[SB]
POP (0) 1
STORE (1) 10[SB]
PUSH 1
LOAD (2) 2[SB]
POP (1) 1
STORE (1) 11[SB]
LOAD (1) 10[SB]
SUBR IOut
LOAD (1) 11[SB]
SUBR IOut
LOAD (2) 2[SB]
POP (0) 1
LOADL 1
SUBR IAdd
LOAD (2) 2[SB]
POP (1) 1
LOADL 2
SUBR IAdd
STORE (2) 2[SB]
LOAD (2) 2[SB]
POP (0) 1
SUBR IOut
LOAD (2) 2[SB]
POP (1) 1
SUBR IOut
LOAD (2) 4[SB]
POP (0) 1
LOADL 2
SUBR IAdd
LOAD (2) 4[SB]
POP (1) 1
LOADL 2
SUBR IAdd
STORE (2) 4[SB]
LOAD (2) 4[SB]
POP (0) 1
SUBR IOut
LOAD (2) 4[SB]
POP (1) 1
SUBR IOut
JUMP end_fonc
fonc
LOAD (2) -2[LB]
RETURN (2) 2
end_fonc
PUSH 2
LOAD (2) 2[SB]
CALL (LB) fonc
STORE (2) 12[SB]
LOAD (2) 12[SB]
POP (0) 1
SUBR IOut
POP (0) 14
HALT

