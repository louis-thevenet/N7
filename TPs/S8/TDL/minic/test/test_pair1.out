true425106122244
